packages/mock-block-dock/src/datastore/traverse.ts

```diff
@@ -0,0 +1,252 @@
+import {
+  EntityEditionId,
+  EntityId,
+  HasLeftEntityEdge,
+  HasRightEntityEdge,
+  IncomingLinkEdge,
+  OntologyTypeEditionId,
+  OutgoingLinkEdge,
+  Subgraph,
+} from "@blockprotocol/graph";
+import {
+  getIncomingLinksForEntity,
+  getLeftEntityForLinkEntity,
+  getOutgoingLinksForEntity,
+} from "@blockprotocol/graph/stdlib";
+import { getRightEntityForLinkEntity } from "@blockprotocol/graph/stdlib-temporal";
+
+import { typedEntries } from "../util";
+import { addKnowledgeGraphEdge } from "./mutate-subgraph";
+import { ResolveMap } from "./traverse/resolve-map";
+import { PartialDepths, TraversalContext } from "./traverse/traversal-context";
+
+/** @todo - doc */
+/** @todo - Update this to take an interval instead of always being "latest" */
+/** @todo - Update this to handle ontology edges */
+export const traverseElement = (
+  traversalSubgraph: Subgraph,
+  elementIdentifier: EntityEditionId | OntologyTypeEditionId,
+  datastore: Subgraph,
+  traversalContext: TraversalContext,
+  currentTraversalDepths: PartialDepths,
+) => {
+  const unresolvedDepths = traversalContext.insert(
+    elementIdentifier,
+    currentTraversalDepths,
+  );
+
+  if (Object.keys(unresolvedDepths).length === 0) {
+    return;
+  }
+
+  const element =
+    datastore.vertices[elementIdentifier.baseId]?.[elementIdentifier.versionId];
+
+  if (!element) {
+    throw new Error(
+      `Couldn't find element in graph associated with identifier: ${JSON.stringify(
+        elementIdentifier,
+      )}`,
+    );
+  }
+
+  const editionsInTraversalSubgraph =
+    traversalSubgraph.vertices[elementIdentifier.baseId];
+
+  // `any` casts here are because TypeScript wants us to narrow the Identifier type before trusting us
+  if (editionsInTraversalSubgraph) {
+    (editionsInTraversalSubgraph as any)[elementIdentifier.versionId] = element;
+  } else {
+    // eslint-disable-next-line no-param-reassign -- The point of this function is to mutate the subgraph
+    (traversalSubgraph as any).vertices[elementIdentifier.baseId] = {
+      [elementIdentifier.versionId]: element,
+    };
+  }
+
+  const toResolve: ResolveMap = new ResolveMap({});
+
+  for (const [edgeKind, depths] of typedEntries(unresolvedDepths)) {
+    // Little hack for typescript, this is wrapped in a function with a return value to get type safety to ensure the
+    // switch statement is exhaustive. Try removing a case to see.
```

Comment:
Unsure if I want to keep this. It makes it harder to read due to the increased indentation but it would be nice if we add support for the other edge kinds this would complain (and did complain when I messed up some stuff while developing).

Extracting it into a standalone function makes it a lot less ergonomic, you have to figure out all the input params, and it'd really need documentation I think.

Reply:
We could do, perhaps in a separate follow-up @kachkaev or someone could look at. 

I'd be slightly concerned at how well eslint might be able to introspect something like this if the union type is generated rather than being explicitly written.